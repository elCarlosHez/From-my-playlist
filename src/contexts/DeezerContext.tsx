import React, { useCallback, useState } from 'react';

interface IDeezerContext {
  token: string;
  setToken: React.Dispatch<React.SetStateAction<string>>;
  fetchDeezer: (url: string) => Promise<any>;
}

export const useDeezerContext = (): IDeezerContext => {
  const [token, setToken] = useState('');

  const fetchDeezer = useCallback(async (url:string): Promise<any> => {
    try {
      // Consider Dezzer has a CORS block. This method won't work in localhost
      const promise = await fetch(`${url}?access_token=${token}`);
      return await promise.json();
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log('Deezer fetch Error: ', error);
      return null;
    }
  }, [token]);

  return { token, setToken, fetchDeezer };
};
