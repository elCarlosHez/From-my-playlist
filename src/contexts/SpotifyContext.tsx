import React, { createContext, useContext, useState } from 'react';

interface ISpotifyContext {
  token: string;
  setToken: (value: string) => void;
  fetchSpotify: <T>(url: string) => Promise<T | null>;
}

const SpotifyContext = createContext<ISpotifyContext | undefined>(undefined);

interface ISpotifyProvider {
  children: JSX.Element;
}

export const SpotifyProvider = (props: ISpotifyProvider): JSX.Element => {
  const { children } = props;
  const [token, setToken] = useState('');

  async function fetchSpotify <T>(url: string): Promise<T | null> {
    try {
      const promise = await fetch(url, {
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
      });
      return await promise.json();
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(error);
      return null;
    }
  }

  return (
    <SpotifyContext.Provider
      // eslint-disable-next-line react/jsx-no-constructed-context-values
      value={{
        token,
        setToken,
        fetchSpotify,
      }}
    >
      {children}
    </SpotifyContext.Provider>
  );
};

export const useSpotifyAuth = () => {
  const context = useContext(SpotifyContext);
  if (context === undefined) {
    throw new Error('useSpotifyAuth must be used within a SpotifyProvider');
  }

  return context;
};
