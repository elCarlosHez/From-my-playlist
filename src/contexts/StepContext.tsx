import { useState } from "react";

export enum Steps {
  start,
  destiny,
  transfer,
  finish
}

export interface ReturnStepContext {
  step: Steps;
  setStep: React.Dispatch<React.SetStateAction<Steps>>;
}

export const stepContext = (): ReturnStepContext => {
  const [step, setStep] = useState<Steps>(Steps.start);

  return {
    step,
    setStep,
  }
};
