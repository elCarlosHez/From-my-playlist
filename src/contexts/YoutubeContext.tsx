import React, { useCallback, useState } from "react";

interface ISpotifyProvider {
  token: string;
  setToken: React.Dispatch<React.SetStateAction<string>>;
  fetchYoutube: (url: string) => Promise<any>;
}

export const useYoutubeContext = (): ISpotifyProvider => {
  const [token, setToken] = useState('');

  const fetchYoutube = useCallback(
    async (url: string): Promise<any> => {
      try {
        const promise = await fetch(`${url}?part=snippet&mine=true`, {
          headers: {
            Accept: "application/json",
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
        });
        return await promise.json();
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error(error);
        return null;
      }
    },
    [token]
  );

  return { token, setToken, fetchYoutube };
};
