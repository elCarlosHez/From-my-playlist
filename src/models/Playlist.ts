export type PlaylistQuery = {
  href: string;
  items: Playlist[];
  limit: number;
}

export interface Playlist {
  id: string;
  name: string;
  href: string;
}