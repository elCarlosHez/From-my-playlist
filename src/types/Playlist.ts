export type Playlist = {
  id: string;
  name: string;
  href: string;
}
