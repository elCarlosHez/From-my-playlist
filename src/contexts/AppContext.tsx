import React, { createContext, useContext, useState } from "react";
import { useDeezerContext } from "./DeezerContext";
import { useSpotifyContext } from "./SpotifyContext";
import { useYoutubeContext } from "./YoutubeContext";

interface IAppContext {
  // Spotify Context
  spotifyToken: string;
  setSpotifyToken: React.Dispatch<React.SetStateAction<string>>;
  fetchSpotify: (url: string) => Promise<any>;
  // Youtube Music Context
  youtubeToken: string;
  setYoutubeToken: React.Dispatch<React.SetStateAction<string>>;
  fetchYoutube: (url: string) => Promise<any>;
}

const AppContext = createContext<IAppContext | undefined>(undefined);

interface IAppProvider {
  children: JSX.Element;
}

export const AppProvider = (props: IAppProvider): JSX.Element => {
  const { children } = props;
  const {
    token: spotifyToken,
    setToken: setSpotifyToken,
    fetchSpotify,
  } = useSpotifyContext();

  const {
    token: youtubeToken,
    setToken: setYoutubeToken,
    fetchYoutube,
  } = useYoutubeContext();

  return (
    <AppContext.Provider
      value={{
        spotifyToken,
        setSpotifyToken,
        fetchSpotify,

        youtubeToken,
        setYoutubeToken,
        fetchYoutube,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = () => {
  const context = useContext(AppContext);
  if (context === undefined) {
    throw new Error("useAppContext must be used within a AppProvider");
  }

  return context;
};
