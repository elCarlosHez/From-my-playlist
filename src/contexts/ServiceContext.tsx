import { useState } from "react";
import { Playlist } from "../types/Playlist";
import { ServicesList } from "../types/ServicesList";

export interface ReturnServiceContext {
  service: ServicesList | null;
  setService: React.Dispatch<React.SetStateAction<ServicesList | null>>
  playlist: Playlist | null;  
  setPlaylist: React.Dispatch<React.SetStateAction<Playlist | null>>;
}

export const serviceContext = (): ReturnServiceContext => {
  const [service, setService] = useState<ServicesList | null>(null);
  const [playlist, setPlaylist] = useState<Playlist | null>(null);

  return {
    service,
    setService,
    playlist,
    setPlaylist
  }
};
