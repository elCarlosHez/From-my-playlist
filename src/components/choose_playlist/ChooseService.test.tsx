import { create } from "react-test-renderer";
import { AppProvider } from "../../contexts/AppContext";
import { Playlist } from "../../types/Playlist";
import { User } from "../../types/User";
import { ChoosePlaylist } from "./";

const playlists: Playlist[] = [{
  id: '1',
  name: 'test',
  image: 'https://dummy.image',
  href: '#',
  author: {
    id: '1',
    name: 'name',
  } as User,
}];

describe("<ChoosePlaylist /> tests", () => {
  test("Render the component wihout crash", () => {
    const component = create(
      <AppProvider>
        <ChoosePlaylist playlists={playlists} />
      </AppProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
