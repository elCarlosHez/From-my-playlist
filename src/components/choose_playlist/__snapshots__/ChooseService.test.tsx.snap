// Vitest Snapshot v1

exports[`<ChoosePlaylist /> tests > Render the component wihout crash 1`] = `
<ul
  className="p-0 flex flex-col mt-24 overflow-y-auto"
>
  <button
    className="grid grid-cols-12 mb-5 border-2 border-gray-300 rounded"
    onClick={[Function]}
  >
    <img
      className="col-span-4"
      src="https://dummy.image"
    />
    <div
      className="col-start-6 col-span-7 text-left w-full h-full flex flex-col items-start justify-center"
    >
      <p
        className="font-bold"
      >
        test
      </p>
      <p
        className="text-sm"
      >
        By name
      </p>
    </div>
  </button>
</ul>
`;
